### PR TITLE
Remove DefaultSpan from public API.

### LINKED ISSUE #1704: Update DefaultSpan to match the new spec
This might just be a rename to PropagatedSpan, but please check the changes to see if there are any behavioral changes as well:  https://github.com/open-telemetry/opentelemetry-specification/pull/994

### PR DESCRIPTION
**Pull request by anuraaga**
We use `DefaultSpan`, a no-op wrapper of a spancontext, for three use cases

- propagated span
- unsampled span
- disabling implicit tracing in a scope

So it actually didn't make sense to use the propagated span terminology to replace it. I used a `wrapAsNoop` factory instead since that's literally what we're doing. ~~To give the propagated term some love, I renamed the relevant `SpanContext` factory to use the word.~~ Realized I was confusing propagated span with remote span but they are different.

Would be much easier to reason about without the `SpanContext` I guess 😋 

Fixes #1704

### CODE REVIEW

File: api/src/main/java/io/opentelemetry/trace/Span.java [hunk 1]
```diff
@@ -23,6 +23,26 @@
...
+  static Span getInvalid() {
+    return DefaultSpan.INVALID;
+  }
+
+  /**
+   * Returns a {@link Span} that only wraps the provided {@link SpanContext} but has no
+   * functionality. It will not be exported and all operations are no-op, but it can be used to
+   * propagate the {@link SpanContext} downstream.
+   */
+  static Span wrapWithNoop(SpanContext spanContext) {
```
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/Span.java:39]: @Oberon00 Based on the conversation in https://github.com/open-telemetry/opentelemetry-specification/issues/1086 this should be something like `wrapForPropagation`, we're wrapping this `SpanContext` into a `PropagatedSpan`. This seems like odd naming to me though, as there is nothing about propagation happening here really, doubly-so because propagation is still more commonly used for remote propagation than in-process propagation I think. That's not great here since we don't do remote propagation of the invalid span.

Comment (Oberon00) [reply api/src/main/java/io/opentelemetry/trace/Span.java:39]: >  as there is nothing about propagation happening here really

If one considers making a span a local parent "propagation" (which I think one should), then it makes sense. Because what other uses would this returned object have?

Comment (anuraaga) [reply api/src/main/java/io/opentelemetry/trace/Span.java:39]: > Because what other uses would this returned object have?

I dunno maybe someone will have something creative :) While it's indirect usage of this API, one case where it will be used is when extracting the `SpanContext` from a link - the result of this method will be taken out of the context, the link will be made, and the result will be dropped without ever being propagated. So this is where the disconnect with the naming comes from - since it's completely up to the caller on whether it's propagated or not, so I find it pretty confusing. There's the other point where there's no such thing as a non-propagated span - all spans can be propagated, so why do we have the propagated span called out in particular? I'm sure someone is going to call `Span.getPropagated(tracer.startSpan().getContext())` assuming that's how you're supposed to propagate your span ;) `Span.getNoop(tracer.startSpan().getContext())`, `Span.getNonRecording(...)` seems more obviously wrong.

Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/Span.java:39]: I don't love this name, but I 100% agree with the sentiment. Making it super clear what the capabilities (and lack thereof) will be of the returned Span seems like a very important signal to the user.

File: api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java [hunk 2]
```diff
@@ -144,44 +144,45 @@ public static HttpTraceContext getInstance() {
...
+    if (!span.isValid()) {
       return context;
     }
 
-    return TracingContextUtils.withSpan(DefaultSpan.create(spanContext), context);
+    return TracingContextUtils.withSpan(span, context);
   }
 
-  private static <C> SpanContext extractImpl(C carrier, Getter<C> getter) {
+  private static <C> Span extractImpl(C carrier, Getter<C> getter) {
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java:155]: This change seems a bit orthogonal to the PR. Does it need to be included?

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/AwsXRayPropagator.java [hunk 3]
```diff
@@ -123,18 +122,18 @@ public static AwsXRayPropagator getInstance() {
     Objects.requireNonNull(carrier, "carrier");
     Objects.requireNonNull(getter, "getter");
 
-    SpanContext spanContext = getSpanContextFromHeader(carrier, getter);
-    if (!spanContext.isValid()) {
+    Span span = extractSpan(carrier, getter);
```
Comment (jkwatson) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/AwsXRayPropagator.java:125]: I'd like to separate out this changes into a follow-on PR, just to keep this one a little more tight in its scope.

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/B3PropagatorExtractorMultipleHeaders.java [hunk 4]
```diff
@@ -29,33 +28,33 @@
...
+    if (!span.isValid()) {
       return Optional.empty();
     }
 
-    return Optional.of(TracingContextUtils.withSpan(DefaultSpan.create(spanContext), context));
+    return Optional.of(TracingContextUtils.withSpan(span, context));
   }
 
-  private static <C> SpanContext getSpanContextFromMultipleHeaders(
+  private static <C> Span extractFromMultipleHeaders(
```
Comment (jkwatson) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/B3PropagatorExtractorMultipleHeaders.java:39]: ditto on reducing scope for this PR

**Comment by anuraaga** [comment]
@jkwatson Good call - I also went ahead and named the method `getPropagated` to stick with the spec for now. Hoping will be able to rename it later, it's just 10secs in IntelliJ to rename :)

**Comment by anuraaga** [comment]
Had a scary unexpected force push but branch seems to be ok, phew

File: api/src/main/java/io/opentelemetry/trace/DefaultTracer.java [hunk 5]
```diff
@@ -64,9 +64,7 @@ public Span startSpan() {
         spanContext = TracingContextUtils.getCurrentSpan().getContext();
       }
 
-      return spanContext != null && !SpanContext.getInvalid().equals(spanContext)
-          ? new DefaultSpan(spanContext)
-          : DefaultSpan.getInvalid();
+      return Span.getPropagated(spanContext);
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/trace/DefaultTracer.java:67]: this is a nice cleanup. much tighter and clearer.

File: api/src/main/java/io/opentelemetry/trace/Span.java [hunk 6]
```diff
@@ -23,6 +23,26 @@
...
+  static Span getInvalid() {
+    return PropagatedSpan.INVALID;
+  }
+
+  /**
+   * Returns a non-recording {@link Span} that holds the provided {@link SpanContext} but has no
+   * functionality. It will not be exported and all tracing operations are no-op, but it can be used
+   * to propagate a valid {@link SpanContext} downstream.
+   */
+  static Span getPropagated(SpanContext spanContext) {
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/Span.java:39]: s/getPropagated/toPropagated?

File: api/src/main/java/io/opentelemetry/trace/Span.java [hunk 7]
```diff
@@ -304,6 +324,15 @@ default void setAttribute(AttributeKey<Long> key, int value) {
...
   boolean isRecording();
 
+  /**
+   * Returns whether this {@link Span} is valid.
+   *
+   * @see Span#getInvalid()
+   */
+  default boolean isValid() {
+    return getContext().isValid();
+  }
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/Span.java:334]: I think this change is unrelated with the PR description and purpose, consider to not add "secondary" changes to a PR.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
I like the overall change of "PropagationSpan" but I don't like that we add other changes to the API like the "isValid" in the same PR. That change is independent and can be added in a separate PR.

File: api/src/main/java/io/opentelemetry/trace/Span.java (hunk 6 above)
Comment (Oberon00) [reply api/src/main/java/io/opentelemetry/trace/Span.java:39]: I think this should be called `Span.wrap(SpanContext)`. Then we are also safe against renamings of propagated/noop/etc

Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/trace/Span.java:39]: I don't think we are safe against behavior change, this needs to always propagate the SC. But +1 for the name

File: api/src/main/java/io/opentelemetry/trace/Span.java (hunk 7 above)
Comment (anuraaga) [reply api/src/main/java/io/opentelemetry/trace/Span.java]: Sure makes sense

**Review by bogdandrutu — APPROVED** [review]
